mantidproject/mantid
# Load instrument in background when opening instrument viewer

**Description of work.**

This allows the instrument to be loaded in a background thread when the instrument viewer is opened to let the window be opened immediately. This prevents currently running scripts from stalling and/or hanging when the instrument viewer is opened on workspaces containing large instruments.

An additional bug was fixed that caused a runtime error after the instrument viewer was opened on the same workspace twice, and then the user deletes the workspace.



**To test:**



Load several different instruments and open the instrument viewer. Verify that the instrument loading is handled in the background, and that various other conditions work as expected (closing the window, replacing the workspace, opening multiple instrument viewer windows, etc).

Open the instrument viewer with a large instrument (like MANDI) while a script is executing. The instrument viewer window should appear instantly rather, and the workbench script should continue to execute and update progress instead of hanging.

Verify that all of the instrument viewer's features still work as expected.

Fixes [SCD402](https://code.ornl.gov/sns-hfir-scse/diffraction/single-crystal/single-crystal-diffraction/-/issues/402). 

Version into `ornl-next`: #32811








---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
